qep-345-line-symbollayer-blankareas.md

```diff
@@ -0,0 +1,117 @@
+# QGIS Enhancement: Markers/Hashes line symbol layer blank areas 
+
+**Date** 2025/07/22
+
+**Author** Julien Cabieces (@troopa81)
+
+**Contact** julien dot cabieces at oslandia dot com
+
+**Version** QGIS 4.0
+
+# Summary
+
+This proposal relates to the line symbol layer which draw markers or hashes at intervals along a line. When using such symbol layer, the display of the repeated pattern (hash or marker) can be ugly, and user would like to have a better control on what's displayed.
+
+![issue.png](images/qep345/issue.png)
+
+The rest of the proposal will mention markers but everything describe here applies also to hashes.
+
+## Proposed Solution
+
+In order to have a proper display of the markers along the line, I propose to:
+- Give user a way to identify blank area where markers are not displayed
+- Use the already existing annotation feature to add, move, rotate or even change marker symbology
+
+This could look like approximatively this
+
+![poc.gif](images/qep345/poc.gif)
+
+### Fixed marker position
+
+We also propose to add an option "fixed marker position" to fix markers positions whatever the current displayed area and zoom level. Indeed, without this option, you would get this behavior
+
+![nonfixed.gif](images/qep345/nonfixed.gif)
+
+This option will be false by default which is the actual behavior.
+
+Enabling this option will imply that we disable feature clipping for the given line symbol layer. The flag `Qgis::SymbolLayerFlag::DisableFeatureClipping` would be added to current flags if this option is set.
+
+When set, all the geometry would be drawn entirely and not clipped. This would have performance implication. A tooltip would inform user about this.
+
+### The blank areas
+
+These areas would be defined as start and end distances from the first point of the line or polygon. It would be possible to define the unit of those distances.
+
+A new property would be added on QgsSymbolLayer [ones](https://github.com/qgis/QGIS/blob/d01be3798eb05c74a61629042f7189ad9c594b1e/src/core/symbology/qgssymbollayer.h#L147) and it would be possible to explicity set a data defined property for all `QgsTemplatedLineSymbolLayerBase` symbol layers (which marker line and hash line symbol layer inherits from). It would be possible then to define those blank areas either:
```

Comment:
what property type would this be? a string value? If so, how do you propose to store them?  (eg comma separated ranges, ...?)

Reply:
I'm still hesitating here between
- string comma separated
- add a new QgsProperty DataTypeNumericList in this [enum](https://github.com/qgis/QGIS/blob/a8f44fa00e7daa3c8c73d098d602c16ff76051ba/src/core/qgsproperty.h#L82)

With a preference for the second one. But I wanted to wait implementation to see what would be best. Do you have any opinion on this ?

Reply:
I think list is theoretically better... but then we'd have patchy format support for it (ie not shapefiles). So unless there's fallback logic to use strings if the layer doesn't support lists, then I'd just play it safe and use strings alone.

Reply:
OK, let's stick for string for now and I'll see if I list can be done (and if it's a good idea) during implementation